content/articles/spring-boot-recaptcha-v3/index.md

```diff
@@ -0,0 +1,855 @@
+    @ManyToMany(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
+    private Collection<Role> roles = new ArrayList<>();
+
+    public AppUser(String userName, String password) {
+        this.userName = userName;
+        this.password = password;
+    }
+}
+```
+Create a class named ReCaptchaResponse with field `success`, `hostname`, `action`, `score`, `challenge_ts`, and `errorCodes` which is an array of string and annotate it with `@JsonProperty("error-codes")` indicating that response from the server is a `JSON` and will be mapped to an array of strings. 
+
+```java
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Data;
+
+@Data
+public class ReCaptchaResponse {
+    private boolean success;
+    private String hostname;
+    private String action;
+    private float score;
+    private String challenge_ts;
+
+    @JsonProperty("error-codes")
+    private String[] errorCodes;
+}
+```
+
+### Create User and Role repositories
+Create an interface named `AppUserRepository` and extend `JpaRepository` which will provide us the capability to perform CRUD operations without writing any queries.
+
+Create a method that returns an `AppUser` by searching for a username and this can be achieved by adding the following method which is created by chaining the field names using special keywords.
+
+```java
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface AppUserRepository extends JpaRepository<AppUser, Long> {
+    AppUser findAppUserByUserName(String username);
+}
+```
+
+Create an interface named `RoleRepository ` as we did previously and add a method that returns a role by searching for a role name.
+
+```java
+import org.springframework.data.jpa.repository.JpaRepository;
+
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Role findRoleByRoleName(String roleName);
+}
+```
+### Add a `CommandLineRunner` to create custom users
+In your main class where the main method is located, create a password encoder `Bean` that will be used to encode the password before saving it to the database.
+
+```java
+@Bean
+PasswordEncoder passwordEncoder(){
+    return new BCryptPasswordEncoder();
+}
+```
+
+Inside the main class, create a `CommandLineRunner` Bean that we will use to create some custom users to test our application when the application bootstraps.
+
+The command-line runner will accept three parameters including `AppUserRepository`, `RoleRepository`, and `PasswordEncoder`.
+
+```java
+@Bean
+    CommandLineRunner commandLineRunner(AppUserRepository appUserRepository,
+                                        RoleRepository roleRepository,
+                                        PasswordEncoder passwordEncoder){
+        return args -> {
+
+        }
+                                       
+}
+```
+
+Inside the arrow function of the command, line runner create a list of roles and use the `RoleRepository` to save the roles in the database using the `saveAll(roles)` method.
```

Comment:
Change to "Inside the command's arrow function, the command-line runner creates a list of roles and uses the `RoleRepository` to save the roles in the database using the `saveAll(roles)` method."